### ISSUE TITLE
Setting maxzoom on a raster-dem source fails, preventing over-zoom

### ISSUE THREAD

**Issue by JeffreyEarly** [mapbox/mapbox-gl-js#9159]
We created a custom raster-dem for sea-surface height topography up to zoom level 4. However, we found that setting a "maxzoom: 4" completely stops mapbox from loading the tiles at all.

Specifically, if you call
```javascript
map.addSource("ssh_dem",{
     type: 'raster-dem',
     tiles: ['https://ssh...'],
     tileSize: 256
});
```
and add a hillshade layer, then the dem will load okay if you're at the appropriate zoom. It will also overzoom *if* the tile was previously loaded.

However, if you set a max zoom on the source,
```javascript
map.addSource("ssh_dem",{
     type: 'raster-dem',
     tiles: ['https://ssh...'],
     tileSize: 256,
     maxzoom: 4
});
```
the whole thing will silently fail.

As a workaround, *not* setting the max zoom doesn't work for us because often the user will already be overzoomed, so the tiles simply won't load.

Here are live examples,

- [without maxzoom](http://ec2-18-191-206-121.us-east-2.compute.amazonaws.com/raster-dem-test-without-maxzoom.php)
- [with maxzoom](http://ec2-18-191-206-121.us-east-2.compute.amazonaws.com/raster-dem-test-with-maxzoom.php)

Any help on this would be greatly appreciated.

**Comment by andrewharvey** [comment]
Could you share a minimal reproducable example please? I just tested this with the example at https://docs.mapbox.com/mapbox-gl-js/example/hillshade/ by adding a maxzoom at the source and it's working as expected, loading up to the maxzoom and overzooming beyond that.

**Comment by JeffreyEarly** [comment]
Strange. I tried that exact sample code, added maxzoom, and it reproduced the bug perfectly.

I've tried this on various Mac running 10.15 with both Safari and Chrome. What are you trying this on?

**Comment by andrewharvey** [comment]
I don't think it's a browser issue. This is what happens when I change the maxzoom on the source to around 18 then 9 then 7. I have no idea why setting maxzoom: 18 and maxzoom: 10 shows different results it should be the same, so looks like there is an issue but this is different to what you're describing. But when you set it less like zoom 7 I can see a more blurred hillshade which is what you'd expect since it's using lower resolution data.

![Screenshot from 2020-01-05 16-14-15](https://user-images.githubusercontent.com/117278/71775546-b1294480-2fd6-11ea-9500-181d2cdbab01.png)
![Screenshot from 2020-01-05 16-13-25](https://user-images.githubusercontent.com/117278/71775547-b1c1db00-2fd6-11ea-8010-0d82f748994f.png)
![Screenshot from 2020-01-05 16-13-06](https://user-images.githubusercontent.com/117278/71775548-b1c1db00-2fd6-11ea-90f1-38c177a5db7a.png)

What do you see?

**Comment by andrewharvey** [comment]
...on second though it seems the maxzoom is affecting the intensity/scaling, but it shouldn't it should be the zoom the tiles come from that affects it, not the maxzoom. This could be the issue. It could be that setting to 4 has such little intensity that you can't see it.

**Comment by JeffreyEarly** [comment]
Yes, I think you're right. If you click on my original examples, you can actually see extremely faint artifacts in the maxzoom case. So maybe it's rendering something, it's just no longer visible.

**Comment by andrewharvey** [comment]
I still think it's a bug, changing the maxzoom on the source shouldn't change how the layer renders for zooms under the maxzoom.

**Comment by JeffreyEarly** [comment]
The relevant line of code appears to be [line 63](https://github.com/mapbox/mapbox-gl-js/blob/master/src/shaders/hillshade_prepare.fragment.glsl) of hillshade_prepare,
```javascript
    vec2 deriv = vec2(
        (c + f + f + i) - (a + d + d + g),
        (g + h + h + i) - (a + b + b + c)
    ) /  pow(2.0, (u_zoom - u_maxzoom) * exaggeration + 19.2562 - u_zoom);
```
which comes with a bunch of good discussion around it.

If you simply fix `u_maxzoom` to something like 14, then the problem goes away,
```javascript
    ) /  pow(2.0, (u_zoom - 14) * exaggeration + 19.2562 - u_zoom);
```
Looking at the [original discussion](https://github.com/mapbox/mapbox-gl-js/pull/5286#discussion_r148419556), with @nickidlugash and @mollymerp, I think it makes sense to fix this value as a "constant" in the problem.

**Comment by JeffreyEarly** [comment]
Any thoughts on how to do a long-term fix on this?

For our own website I ended up hard coding u_maxzoom=22.0 into hillshade_prepare because 22 matches the default assumed maxzoom level for a source. But it would be nice to get a long term fix.

It might be helpful for the discussion to know what our custom terrain tiles are. We have sea-surface height (SSH) topography encoded into terrain-rgb tiles. In our case though, because the topographic relief is typically a maximum of +/- 80 cm depending on the day, I ended up multiplying by 10000, so typical heights match the largest mountains in the world. The hillshade algorithm is then happy.
